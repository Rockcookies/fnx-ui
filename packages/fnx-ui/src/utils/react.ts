import {
	Children,
	forwardRef,
	ForwardRefExoticComponent,
	ForwardRefRenderFunction,
	FunctionComponent,
	PropsWithoutRef,
	ReactElement,
	RefAttributes,
} from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};

export const createForwardRef = <R, P = any>(
	displayName: string,
	render: ForwardRefRenderFunction<R, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<R>> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsForwardRef = <R, P, D>(
	displayName: string,
	defaultProps: D,
	render: ForwardRefRenderFunction<R, P & D>,
): ForwardRefExoticComponent<
	PropsWithoutRef<P & Partial<D>> & RefAttributes<R>
> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};

export const createFC = <P = any>(
	displayName: string,
	fc: FunctionComponent<P>,
): FunctionComponent<P> => {
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsFC = <P, D>(
	displayName: string,
	defaultProps: D,
	fc: FunctionComponent<P & D>,
): FunctionComponent<P & Partial<D>> => {
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};
