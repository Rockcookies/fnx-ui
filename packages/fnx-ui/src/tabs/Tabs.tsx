import React, {
	useContext,
	useEffect,
	useImperativeHandle,
	useMemo,
	useRef,
	useState,
} from 'react';
import { TabsRef } from '.';
import ConfigProvider from '../config-provider';
import configComponentProps from '../hooks/config-component-props';
import useControlledState from '../hooks/use-controlled-state';
import useCreation from '../hooks/use-creation';
import useDefaults from '../hooks/use-defaults';
import Sticky from '../sticky';
import Swipe from '../swipe';
import { SwipeRef } from '../swipe/interface';
import { clamp } from '../utils/format';
import { classnames } from '../utils/namespace';
import { createForwardRef } from '../utils/react';
import { TabsContext } from './context';
import useTabMetaList from './hooks/use-tab-meta-list';
import { TabsProps } from './interface';
import TabsHeader, { TabsHeaderRef } from './TabsHeader';
import { _bem as bem } from './utils';

const useProps = configComponentProps<
	Required<
		Pick<
			TabsProps,
			| 'border'
			| 'animated'
			| 'ellipsis'
			| 'sticky'
			| 'stickyProps'
			| 'tabSwipeThreshold'
			| 'swipeable'
			| 'slots'
		>
	>
>({
	border: false,
	animated: false,
	ellipsis: true,
	sticky: false,
	stickyProps: {},
	tabSwipeThreshold: 4,
	swipeable: false,
	slots: {},
});

const Tabs = createForwardRef<TabsRef, TabsProps>('Tabs', (_props, ref) => {
	const rootRef = useRef<HTMLDivElement>(null);
	const swipeRef = useRef<SwipeRef>(null);
	const tabsHeaderRef = useRef<TabsHeaderRef>(null);

	const [
		{
			border,
			animated,
			ellipsis,
			sticky,
			stickyProps,
			tabSwipeThreshold,
			swipeable,
			slots,
		},
		{
			defaultActiveKey,
			activeKey: _activeKey,
			onChange: _onChange,
			onTabClick,
			trackWidth,
			trackHeight,
			trackColor,
			transitionDuration: _transitionDuration,
			children,
			className,
			...restProps
		},
	] = useProps(_props);

	useImperativeHandle<TabsRef, TabsRef>(
		ref,
		() => {
			return {
				root: rootRef.current,
				reset: () => {
					tabsHeaderRef.current?.reset();
					swipeRef.current?.reset();
				},
			};
		},
		[tabsHeaderRef],
	);

	const configContext = useContext(ConfigProvider.Context);

	const transitionDuration = useDefaults<number>(
		configContext.transitionDuration,
		_transitionDuration,
	);

	const { value: activeKey, onChangeRef } = useControlledState<
		string | undefined
	>({
		value: _activeKey != null ? String(_activeKey) : undefined,
		defaultValue: () =>
			defaultActiveKey != null ? String(defaultActiveKey) : undefined,
		onChange:
			_onChange != null
				? (v) => {
						if (v != null) {
							_onChange(v);
						}
				  }
				: undefined,
	});

	const [metaKeyList, metaList] = useTabMetaList(children);

	const activeIndex = useCreation<number>(
		(prev) => {
			const next = metaKeyList.findIndex((key) => key === activeKey);

			return clamp(
				next < 0 ? prev || 0 : next,
				0,
				Math.max(metaKeyList.length - 1, 0),
			);
		},
		[metaKeyList, activeKey],
	);

	const [swipeActiveIndex, setSwipeActiveIndex] =
		useState<number>(activeIndex);

	// whether the tabList is scrollable
	const scrollable = useMemo(() => {
		return metaList.length > tabSwipeThreshold || !ellipsis;
	}, [metaList.length, tabSwipeThreshold, ellipsis]);

	useEffect(() => {
		if (swipeRef.current && swipeActiveIndex !== activeIndex) {
			swipeRef.current.swipeTo(activeIndex, { animation: true });
			setSwipeActiveIndex(activeIndex);
		}
	}, [activeIndex, swipeActiveIndex]);

	const renderHeader = () => {
		return (
			<TabsHeader
				transitionDuration={transitionDuration}
				activeIndex={activeIndex}
				scrollable={scrollable}
				ellipsis={ellipsis}
				slots={slots}
				metaList={metaList}
				trackWidth={trackWidth}
				trackHeight={trackHeight}
				trackColor={trackColor}
				border={border}
				onChange={(key) => {
					onChangeRef.current(key);
				}}
				onTabClick={onTabClick}
				ref={tabsHeaderRef}
			/>
		);
	};

	const renderChildren = () => {
		if (animated || swipeable) {
			return (
				<Swipe
					defaultActiveIndex={activeIndex}
					className={bem('swipe')}
					loop={false}
					autoplay={0}
					duration={transitionDuration}
					touchable={swipeable}
					indicator={false}
					onActiveIndexChange={(idx) => {
						onChangeRef.current(metaList[idx].key);
						setSwipeActiveIndex(idx);
					}}
					ref={swipeRef}
				>
					{metaList.map((meta) => meta.panel)}
				</Swipe>
			);
		}

		return metaList.map((meta) => meta.panel);
	};

	return (
		<div
			className={classnames(bem(), className)}
			{...restProps}
			ref={rootRef}
		>
			{sticky ? (
				<Sticky {...stickyProps} container={() => rootRef.current}>
					{renderHeader()}
				</Sticky>
			) : (
				renderHeader()
			)}

			<div className={bem('content', { animated })}>
				<TabsContext.Provider
					value={{
						activeKey:
							metaList[activeIndex] && metaList[activeIndex].key,
						animated,
						swipeable,
					}}
				>
					{renderChildren()}
				</TabsContext.Provider>
			</div>
		</div>
	);
});

export default Tabs;
