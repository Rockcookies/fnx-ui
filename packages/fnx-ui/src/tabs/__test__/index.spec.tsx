import { act, fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import Tabs from '../index';

describe('<Tabs/>', () => {
	const mockRect = (el: Element, rect: Partial<DOMRect>) => {
		Object.defineProperty(el, 'getBoundingClientRect', {
			configurable: true,
			get: () => {
				return () => ({
					top: 0,
					left: 0,
					right: 0,
					bottom: 0,
					width: 0,
					height: 0,
					...rect,
				});
			},
		});
	};

	const mockSticky = (sticky: Element, rect?: Partial<DOMRect>) => {
		Object.defineProperty(sticky, 'offsetParent', {
			configurable: true,
			get: () => document.body,
		});

		if (rect) {
			mockRect(sticky, rect);
		}
	};

	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		jest.useRealTimers();
	});

	it('border props', () => {
		const { container } = render(
			<Tabs border={true}>
				<Tabs.Panel title="1">content1</Tabs.Panel>
				<Tabs.Panel title="2">content2</Tabs.Panel>
			</Tabs>,
		);

		expect(
			container.querySelector('.fnx-hairline--top-bottom'),
		).not.toBeNull();
	});

	/* it('trackWidth、trackHeight、trackColor props', async () => {
		const { container } = render(
			<Tabs trackWidth={10} trackHeight={4} trackColor={'green'}>
				<Tabs.Panel title="1">content1</Tabs.Panel>
				<Tabs.Panel title="2">content2</Tabs.Panel>
				<Tabs.Panel title="3">content3</Tabs.Panel>
			</Tabs>,
		);

		await waitFor(() => {
			const track = TestsDOM.mustQuerySelector(
				container,
				'.fnx-tabs__track',
			);
			const style = getComputedStyle(track);

			expect(style.width).toEqual('10px');
			expect(style.height).toEqual('4px');
			expect(style.background).toEqual('green');
		});
	}); */

	it('activeKey、defaultActiveKey', () => {
		const { container, rerender } = render(
			<Tabs defaultActiveKey={'2'}>
				<Tabs.Panel title="1">content1</Tabs.Panel>
				<Tabs.Panel title="2">content2</Tabs.Panel>
				<Tabs.Panel title="3">content3</Tabs.Panel>
			</Tabs>,
		);

		const tabs = container.querySelectorAll('.fnx-tabs__tab');
		expect(
			tabs[2].classList.contains('fnx-tabs__tab--active'),
		).toBeTruthy();

		rerender(
			<Tabs defaultActiveKey={'2'} activeKey={'1'}>
				<Tabs.Panel title="1">content1</Tabs.Panel>
				<Tabs.Panel title="2">content2</Tabs.Panel>
				<Tabs.Panel title="3">content3</Tabs.Panel>
			</Tabs>,
		);
		expect(
			tabs[1].classList.contains('fnx-tabs__tab--active'),
		).toBeTruthy();
	});

	it('click to switch tab and set key', () => {
		const onChange = jest.fn();

		const { container } = render(
			<Tabs onChange={onChange}>
				<Tabs.Panel title="1" key={1}>
					content1
				</Tabs.Panel>
				<Tabs.Panel title="2" key={2}>
					content2
				</Tabs.Panel>
				<Tabs.Panel title="3" key={3}>
					content3
				</Tabs.Panel>
			</Tabs>,
		);

		const tabs = container.querySelectorAll('.fnx-tabs__tab');
		act(() => {
			fireEvent.click(tabs[0]);
		});
		expect(onChange).not.toHaveBeenCalled();

		act(() => {
			fireEvent.click(tabs[1]);
		});
		expect(onChange).toHaveBeenCalledWith('2');
	});

	it('animated、transitionDuration', async () => {
		const { container } = render(
			<Tabs animated={true} transitionDuration={200}>
				<Tabs.Panel title="1">content1</Tabs.Panel>
				<Tabs.Panel title="2">content2</Tabs.Panel>
				<Tabs.Panel title="3">content3</Tabs.Panel>
			</Tabs>,
		);

		const tabs = container.querySelectorAll('.fnx-tabs__tab');
		const swipeItems = container.querySelectorAll('.fnx-swipe-item');

		act(() => {
			fireEvent.click(tabs[1]);
		});

		expect(
			swipeItems[1].classList.contains('fnx-tabs__swipe-item--inactive'),
		).not.toBeTruthy();
	});

	it('onTabClick event', () => {
		const onTabClick = jest.fn();

		const { container } = render(
			<Tabs onTabClick={onTabClick}>
				<Tabs.Panel title="1">content1</Tabs.Panel>
				<Tabs.Panel title="2" disabled={true}>
					content2
				</Tabs.Panel>
				<Tabs.Panel title="3">content3</Tabs.Panel>
			</Tabs>,
		);

		const tabs = container.querySelectorAll('.fnx-tabs__tab');

		act(() => {
			fireEvent.click(tabs[2]);
		});
		expect(onTabClick).toHaveBeenCalled();
	});

	it('sticky, onStickyChange', async () => {
		const onStickyChange = jest.fn();

		const { container } = render(
			<Tabs
				sticky={true}
				stickyProps={{
					onChange: onStickyChange,
				}}
			>
				<Tabs.Panel title="1">content1</Tabs.Panel>
				<Tabs.Panel title="2" disabled={true}>
					content2
				</Tabs.Panel>
				<Tabs.Panel title="3">content3</Tabs.Panel>
			</Tabs>,
		);

		const sticky = container.querySelector('.fnx-sticky');
		const c = container.querySelector('.fnx-tabs');

		c && mockRect(c, { width: 300, height: 200, top: 0, bottom: 600 });
		sticky &&
			mockSticky(sticky, {
				width: 300,
				height: 50,
				top: -10,
			});

		await waitFor(() => {
			fireEvent.scroll(window);
		});
		expect(onStickyChange).toHaveBeenCalledWith({ fixed: true });

		c && mockRect(c, { width: 300, height: 200, top: 0, bottom: -10 });
		await waitFor(() => {
			fireEvent.scroll(window);
		});
		expect(onStickyChange).toHaveBeenCalledWith({ fixed: false });
	});

	it('slots', () => {
		const { container } = render(
			<Tabs
				slots={{
					navLeft: <div>left</div>,
					navRight: <div>right</div>,
					navBottom: 'bottom',
				}}
			>
				<Tabs.Panel title="1">content1</Tabs.Panel>
			</Tabs>,
		);

		expect(container).toMatchSnapshot();
	});
});
