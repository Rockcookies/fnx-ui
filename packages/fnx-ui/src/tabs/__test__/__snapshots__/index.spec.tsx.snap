// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs/> slots 1`] = `
<div>
  <div
    class="fnx-tabs"
  >
    <div
      class="fnx-tabs__header"
    >
      <div
        class="fnx-tabs__nav"
      >
        <div
          class="fnx-tabs__nav-left"
        >
          <div>
            left
          </div>
        </div>
        <div
          class="fnx-tabs__tab-list"
          role="tablist"
        >
          <div
            aria-selected="true"
            class="fnx-tabs__tab fnx-tabs__tab--active"
            role="tab"
          >
            <div
              class="fnx-tabs__tab-title fnx-tabs__tab-title--ellipsis"
            >
              1
            </div>
          </div>
        </div>
        <div
          class="fnx-tabs__nav-right"
        >
          <div>
            right
          </div>
        </div>
      </div>
      bottom
    </div>
    <div
      class="fnx-tabs__content"
    >
      <div
        class="fnx-tabs__panel"
        data-fnx-tab-panel-key="0"
        role="tabpanel"
      >
        content1
      </div>
    </div>
  </div>
</div>
`;
